ostd/src/task/atomic_mode.rs

```diff
@@ -45,3 +48,39 @@ pub fn might_sleep() {
         );
     }
 }
+
+/// A marker trait for guard types that enforce the atomic mode.
+///
+/// Key kernel primitives such as `SpinLock` and `Rcu` rely on
+/// [the atomic mode](crate::task::atomic_mode).
+/// The existence of such a guard guarantees that
+/// the current task is executing in the atomic mode.
+///
+/// Such guard types include [`DisabledLocalIrqGuard`] and [`DisabledPreemptGuard`].
+/// Users can also implement this trait for additional guard types via
+/// [`BorrowAsAtomicModeGuard`] safely.
+///
+/// # Safety
+///
+/// The implementer must ensure that the atomic mode is maintained while
+/// the guard type is alive.
+pub unsafe trait AtomicModeGuard {}
+
+// SAFETY: The guard disables local IRQs, which meets the first
+// sufficient condition for atomic mode.
+unsafe impl AtomicModeGuard for DisabledLocalIrqGuard {}
+// SAFETY: The guard disables preemptions, which meets the second
+// sufficient condition for atomic mode.
+unsafe impl AtomicModeGuard for DisabledPreemptGuard {}
+
+/// A guard that can be borrowed as an [`AtomicModeGuard`].
+pub trait BorrowAsAtomicModeGuard {
```

Comment:
I think that this trait adds unnecessary complexity. Instead of adding a new trait and implementing it for the guard types, we can instead implement the existing trait `AtomicModeGuard` for those guard types directly. Then, we can pass lock guards (e.g., `RwLockWriteGuard_`) to functions that require atomic mode guards.

Reply:
This is necessary because some guard can be defined outside of OSTD, like `DisableBottomHalfGuard`. I [mention this earlier](https://github.com/asterinas/asterinas/pull/1928#issuecomment-2742164540) so it's the motivation to add the `BorrowedAsAtomicModeGuard` trait.